src/wiring_diagrams/ScheduleUndirected.jl

```diff
@@ -0,0 +1,219 @@
+  outer_junc::FinFunction{Int})
+```
+
+where `values` is a list of computed values for boxes or composites in the
+diagram; `juncs` is an equal-length list of `FinSet` functions, mapping the
+ports of each value to junctions; and `outer_junc` is a `FinSet` function
+mapping the outer ports to junctions. This interface is similar to the `ncon`
+function in the [NCON package](https://arxiv.org/abs/1402.0939) for MATLAB or
+the `ncon` function in
+[TensorOperations.jl](https://github.com/Jutho/TensorOperations.jl), except that
+the outer junctions are represented explictly by a third argument rather than
+implictly by using negative numbers in the second argument.
+
+The length of lists `values` and `juncs` in each call depends on the schedule.
+Most scheduling algorithms will reduce the original UWD to binary composites, so
+that the length is always 2, apart from possible degenerate cases of nullary or
+unary composites.
+
+Nested UWDs are used as an auxiliary data structure during the evaluation. If
+the schedule will be evaluated multiple times, you should explicitly convert it
+to a nested UWD using [`to_nested_diagram`](@ref) and then call `eval_schedule`
+on the resulting object instead.
+"""
+eval_schedule(f, s::AbstractScheduledUWD, generators::AbstractVector) =
+  eval_schedule(f, to_nested_diagram(s), generators)
+
+function eval_schedule(f, d::AbstractNestedUWD, generators::AbstractVector)
+  # Evaluate `f` after normalizing junctions to consecutive numbers 1:n.
+  # Question: Should this normalization be optional? It's convenient but also
+  # has some (small) cost.
+  function do_eval(values, juncs, outer_junc)
+    jmap = Dict{Int,Int}()
+    for j in Iterators.flatten((Iterators.flatten(juncs), outer_junc))
+      get!(jmap, j) do; length(jmap) + 1 end
+    end
+    njunc = length(jmap)
+    f(values, map(js -> FinFunction([jmap[j] for j in js], njunc), juncs),
+      FinFunction([jmap[j] for j in outer_junc], njunc))
+  end
+
+  # Mutually recursively evaluate children of composite `c`.
+  function eval_children(c::Int)
+    bs, cs = box_children(d, c), children(d, c)
+    values = [ generators[bs]; map(eval_composite, cs) ]
+    juncs = [ [junction(d, ports(d, b)) for b in bs];
+              [composite_junction(d, composite_ports(d, c′)) for c′ in cs] ]
+    (values, juncs)
+  end
+
+  # Mutually recursively evaluate composite `c`.
+  function eval_composite(c::Int)
+    values, juncs = eval_children(c)
+    outer_junc = composite_junction(d, composite_ports(d, c))
+    do_eval(values, juncs, outer_junc)
+  end
+
+  # Evaluate diagram starting at the root, assumed to be unique. The root
+  # composite must be handled specially due to outer ports.
+  root = only(filter(c -> parent(d, c) == c, composites(d)))
+  values, juncs = eval_children(root)
+  do_eval(values, juncs, junction(d, outer=true))
+end
+
+""" Convert a scheduled UWD to a nested UWD.
+"""
+function to_nested_diagram(s::AbstractScheduledUWD)
```

Comment:
I know we use `to_graphviz` and `to_hom_expr` so this is consistent, but it has bothered me for a while, we should just make these constructors for their return type. That can be a cleanup PR at some point before Catlab v1.0

Reply:
In some cases I agree, in others not. Anyway, yeah, we can defer that discussion to later.